File: iOS/Source/AUIAICallVoiceprintManager.swift
//
//  AUIAICallVoiceprintManager.swift
//  AUIAICall
//
//  Created by Bingo on 2025/07/04.
//

import UIKit
import AliyunOSSiOS


@objcMembers open class AUIAICallVoiceprintManager: NSObject {
    
    public static let shared = AUIAICallVoiceprintManager()
    
    public override init() {
        super.init()

        _ = self.loadData()
    }
    
    // 获取声纹降噪开关
    open private(set) var isEnable: Bool = true
    
    public private(set) var voiceprintItem: AUIAICallVoiceprintItem? = nil
    
    public var userId: String = "" {
        didSet {
            if self.loadData() == false {
                self.reset()
            }
        }
    }
    
    public var region: String {
#if DEMO_FOR_DEBUG
        AUIAICallDebugManager.shared.getRegion()
#else
        AUIAICallAgentConfig.shared.getRegion()
#endif
    }
    private var appserver: AUIAICallAppServer? = nil
    
    
    // 是否完整声纹录入
    open func isRegistedVoiceprint() -> Bool {
        return self.voiceprintItem != nil
    }
    
    // 通话时，可否可以使用声纹降噪能力
    open func canUseVoiceprint() -> Bool {
        return self.isEnable && self.isRegistedVoiceprint()
    }
    
    // 是否开启声纹降噪
    open func enableVoiceprint(_ isEnable: Bool) -> Void {
        self.isEnable = isEnable
        self.saveData()
    }
    
    public func getDirectory() -> URL {
        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
        let url = paths[0].appendingPathComponent("voiceprint")
        if !FileManager.default.fileExists(atPath: url.path) {
            do {
                try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true, attributes: nil)
            } catch {
                debugPrint("创建文件夹失败：\(url)")
            }
        }
        return url
    }
    
    private func getStorePath(filePath: String) -> String {
        let dir = self.getDirectory().absoluteString
        if filePath.hasPrefix(dir) {
            return String(filePath.suffix(from: dir.endIndex))
        }
        return filePath
    }
    
    private func reset() {
        self.isEnable = true
        self.voiceprintItem = nil
        self.saveData()
    }
    
    private func loadData() -> Bool {
        let fileURL = self.getDirectory().appendingPathComponent("\(self.userId)_data.json")
        do {
            let data = try Data(contentsOf: fileURL)
            if let dict = String.init(data: data, encoding: .utf8)?.aicall_jsonObj() {
                if let isEnable = dict["Enable"] as? Bool {
                    self.isEnable = isEnable
                }
                if let itemDict = dict["Item"] as? [String: AnyHashable] {
                    self.voiceprintItem = AUIAICallVoiceprintItem(dict: itemDict)
                }
            }
            return true
        } catch {
            debugPrint("读取失败: \(error)")
        }
        return false
    }

    private func saveData() {
        let fileURL = self.getDirectory().appendingPathComponent("\(self.userId)_data.json")
        do {
            let dict: [String: AnyHashable] = [
                "Enable": self.isEnable,
                "Item": self.voiceprintItem?.toDict()
            ]
            
            let data = dict.aicall_jsonString.data(using: .utf8)
            try data?.write(to: fileURL)
        } catch {
            debugPrint("存储失败: \(error)")
        }
    }

    private func getOssConfig(completed:((_ config: [String: Any]?,  _ error: NSError?) -> Void)?) {
        self.appserver?.request(path: "/api/v2/aiagent/getOssConfig", body: ["user_id": self.userId]) { response, data, error in
            if error == nil {
                completed?(data as? [String: Any], error)
            }
            else {
                completed?(nil, error)
            }
        }
    }

    var client: OSSClient!
    private func uploadFile(audioFileUrl: URL, info: [String: Any], completed:((_ ossUrl: String?, _ error: NSError?)->Void)?) {
        debugPrint("开始上传：\(audioFileUrl)")
        let accessKeyId = (info["access_key_id"] as? String) ?? ""
        let accessKeySecret = (info["access_key_secret"] as? String) ?? ""
        let securityToken = (info["sts_token"] as? String) ?? ""
        let endpoint = "https://\((info["region"] as? String) ?? "").aliyuncs.com"
        let bucketName = (info["bucket"] as? String) ?? "ai-agent-demo"
        let objectKey = ((info["base_path"] as? String) ?? "voice") + "/" + audioFileUrl.lastPathComponent

        let credentialProvider = OSSStsTokenCredentialProvider(accessKeyId: accessKeyId,
                                                               secretKeyId: accessKeySecret,
                                                               securityToken: securityToken)
        let config = OSSClientConfiguration()
        config.maxRetryCount = 3
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60

        self.client = OSSClient(endpoint: endpoint,
                           credentialProvider: credentialProvider,
                           clientConfiguration: config)

        let putRequest = OSSPutObjectRequest()
        putRequest.bucketName = bucketName
        putRequest.objectKey = objectKey
        putRequest.uploadingFileURL = audioFileUrl

        self.client.putObject(putRequest).continue({ (task) -> AnyObject? in
            if let error = task.error as? NSError {
                debugPrint("上传失败: \(error)")
                DispatchQueue.main.async {
                    completed?(nil, error)
                }
            } else {
                if let ossUrl = self.client.presignConstrainURL(withBucketName: bucketName, withObjectKey: objectKey, withExpirationInterval: 24 * 3600).result as? String {
                    debugPrint("上传成功：\(ossUrl)")
                    DispatchQueue.main.async {
                        completed?(ossUrl, nil)
                    }
                }
                else {
                    DispatchQueue.main.async {
                        completed?(nil, NSError.aicall_create(code: -1, message: "Fetch ossurl failed"))
                    }
                }
            }

            return nil
        })
    }

    private func registerVoiceprint(audioOssUrl: String, completed:((_ voiceId: String?, _ error: NSError?)->Void)?) {
        let voiceprintId = "\(self.userId)_\(String.aicall_random())"
        let input = [
            "Type": "oss",
            "Data": audioOssUrl,
            "Format": "wav",
        ]
        let body: [String: Any] = [
            "user_id": self.userId,
            "region": self.region,
            "voiceprint_id": voiceprintId,
            "input": input.aicall_jsonString
        ]
        self.appserver?.request(path: "/api/v2/aiagent/setAIAgentVoiceprint", body: body) { response, data, error in
            if let data = data {
                if let voiceprintId = data["voiceprint_id"] as? String {
                    completed?(voiceprintId, nil)
                    return
                }
                
                let code = (data["code"] as? Int) ?? -1
                let message = data.aicall_jsonString
                completed?(nil, NSError.aicall_create(code: code, message: message))
                return
            }
            completed?(nil, error ?? NSError.aicall_create(code: -1, message: data?.aicall_jsonString ?? "unknown"))
        }
    }

    public func start(audioFileUrl: URL, completed:((_ error: NSError?)->Void)?) {
        
        if self.userId.isEmpty == true {
            completed?(NSError.aicall_create(code: .InvalidParames, message: "userId is empty"))
            return
        }
        
        if self.appserver == nil {
            self.appserver = AUIAICallAppServer()
        }

        let resultBlock: (_ error: NSError?)->Void = { [weak self] error in
            self?.appserver = nil
            completed?(error)
        }

        self.getOssConfig() {[weak self] config, error in
            guard let self = self else { return }
            if let error = error {
                resultBlock(error)
                return
            }
            guard let info = config else {
                resultBlock(NSError.aicall_create(code: -1, message: "Fetch ossconfig failed"))
                return
            }
            self.uploadFile(audioFileUrl: audioFileUrl, info: info, completed: { [weak self] ossUrl, error in
                guard let self = self else { return }
                if let error = error {
                    resultBlock(error)
                    return
                }
                self.registerVoiceprint(audioOssUrl: ossUrl!) { [weak self] voiceprintId, error in
                    guard let self = self else { return }
                    if let error = error {
                        resultBlock(error)
                        return
                    }
                    if let voiceprintId = voiceprintId {
                        let item = AUIAICallVoiceprintItem(voiceprintId: voiceprintId)
                        item.filePath = self.getStorePath(filePath: audioFileUrl.absoluteString)
                        item.ossUrl = ossUrl!
                        self.voiceprintItem = item
                        self.saveData()
                        resultBlock(nil)
                        return
                    }
                    resultBlock(NSError.aicall_create(code: -1, message: "Register voiceprint failed"))
                }
            })
        }
    }
}

// 声纹信息结构
@objcMembers open class AUIAICallVoiceprintItem: NSObject {
    
    public init(voiceprintId: String) {
        self.voiceprintId = voiceprintId
    }
    
    public private(set) var voiceprintId: String = ""
    public var name: String = ""
    public var filePath: String = ""
    public var ossUrl: String = ""
    
    public init(dict: [String: AnyHashable]) {
        if let voiceprintId = dict["voiceprintId"] as? String {
            self.voiceprintId = voiceprintId
        }
        if let name = dict["name"] as? String {
            self.name = name
        }
        if let filePath = dict["filePath"] as? String {
            self.filePath = filePath
        }
        if let ossUrl = dict["ossUrl"] as? String {
            self.ossUrl = ossUrl
        }
    }
    
    open func toDict() -> [String: AnyHashable] {
        return [
            "voiceprintId": self.voiceprintId,
            "name": self.name,
            "filePath": self.filePath,
            "ossUrl": self.ossUrl,
        ]
    }

}

File: Android/AUIAICall/src/main/java/com/aliyun/auikits/aicall/voiceprint/AUIAICallVoiceprintManager.java
package com.aliyun.auikits.aicall.voiceprint;

import android.content.Context;
import android.text.TextUtils;

import com.alibaba.sdk.android.oss.ClientException;
import com.alibaba.sdk.android.oss.OSSClient;
import com.alibaba.sdk.android.oss.ServiceException;
import com.alibaba.sdk.android.oss.callback.OSSCompletedCallback;
import com.alibaba.sdk.android.oss.common.auth.OSSCredentialProvider;
import com.alibaba.sdk.android.oss.common.auth.OSSStsTokenCredentialProvider;
import com.alibaba.sdk.android.oss.internal.OSSAsyncTask;
import com.alibaba.sdk.android.oss.model.PutObjectRequest;
import com.alibaba.sdk.android.oss.model.PutObjectResult;
import com.aliyun.auikits.aiagent.service.ARTCAICallServiceImpl;
import com.aliyun.auikits.aiagent.service.IARTCAICallService;
import com.aliyun.auikits.aiagent.util.Logger;
import com.aliyun.auikits.aicall.util.AppServiceConst;
import com.aliyun.auikits.aicall.util.AUIAICallAgentDebug;
import com.aliyun.auikits.aicall.util.SettingStorage;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;

/**
 * 声纹管理器，统一管理声纹配置、注册、删除等操作
 */
public class AUIAICallVoiceprintManager {

    private static AUIAICallVoiceprintManager sInstance = new AUIAICallVoiceprintManager();

    public static AUIAICallVoiceprintManager getInstance() {
        return sInstance;
    }

    private Context appContext;
    private ARTCAICallServiceImpl.AppServerService appServerService;

    // 声纹开关
    private boolean isEnable = true;

    // 声纹模式：false 预注册，true 无感注册
    private boolean isAutoRegister = false;

    // 当前用户ID
    private String userId = "";

    // 是否预发环境
    private boolean isPreEnv = false;

    // 预注册声纹数据
    private AUIAICallVoiceprintItem preRegisterVoiceprintItem = null;

    // 无感注册声纹数据
    private AUIAICallVoiceprintItem autoRegisterVoiceprintItem = null;

    // 当前通话使用的声纹 ID（临时缓存，用于无感注册回调）
    private String currentVoiceprintId = null;

    private AUIAICallVoiceprintManager() {
    }

    /**
     * 初始化，传入 Application Context
     */
    public void init(Context context) {
        this.appContext = context.getApplicationContext();
    }

    /**
     * 设置用户ID，会触发数据加载
     */
    public void setUserId(String userId) {
        if (TextUtils.equals(this.userId, userId)) {
            return;
        }
        this.userId = userId;
        if (!loadData()) {
            reset();
        }
    }

    /**
     * 设置环境（预发/线上），会触发数据重新加载
     */
    public void setPreEnv(boolean preEnv) {
        if (this.isPreEnv == preEnv) {
            return;
        }
        this.isPreEnv = preEnv;
        if (!loadData()) {
            reset();
        }
        this.appServerService = null;
    }

    /**
     * 获取声纹区域，固定返回 cn-shanghai
     */
    public String getRegion() {
        return "cn-shanghai";
    }

    /**
     * 获取声纹开关状态
     */
    public boolean isEnable() {
        return isEnable;
    }

    /**
     * 设置声纹开关
     */
    public void enableVoiceprint(boolean enable) {
        this.isEnable = enable;
        saveData();
    }

    /**
     * 获取当前注册模式
     * @return false 预注册，true 无感注册
     */
    public boolean isAutoRegister() {
        return isAutoRegister;
    }

    /**
     * 切换声纹注册模式
     * @param autoRegister false 预注册，true 无感注册
     */
    public void switchVoiceprintMode(boolean autoRegister) {
        this.isAutoRegister = autoRegister;
        saveData();
    }

    /**
     * 判断当前模式下是否已注册声纹
     */
    public boolean isRegistedVoiceprint() {
        if (isAutoRegister) {
            return autoRegisterVoiceprintItem != null;
        }
        return preRegisterVoiceprintItem != null;
    }

    /**
     * 获取当前模式下的声纹ID
     */
    public String getVoiceprintId() {
        if (isAutoRegister) {
            return autoRegisterVoiceprintItem != null ? autoRegisterVoiceprintItem.getVoiceprintId() : null;
        }
        return preRegisterVoiceprintItem != null ? preRegisterVoiceprintItem.getVoiceprintId() : null;
    }

    /**
     * 获取声纹配置文件存储目录
     */
    private File getDirectory() {
        if (appContext == null) {
            return null;
        }
        File dir = new File(appContext.getFilesDir(), "voiceprint");
        if (!dir.exists()) {
            boolean ok = dir.mkdirs();
            if (!ok) {
                Logger.e("Failed to create voiceprint directory: " + dir.getAbsolutePath());
            }
        }
        return dir;
    }

    /**
     * 获取当前用户+环境的声纹配置文件
     */
    private File getDataFile() {
        File dir = getDirectory();
        if (dir == null) {
            return null;
        }
        String suffix = isPreEnv ? "_data_pre.json" : "_data.json";
        return new File(dir, userId + suffix);
    }

    /**
     * 重置所有声纹状态
     */
    private void reset() {
        isEnable = true;
        isAutoRegister = false;
        preRegisterVoiceprintItem = null;
        autoRegisterVoiceprintItem = null;
    }

    /**
     * 从 JSON 文件加载声纹配置
     */
    private boolean loadData() {
        if (TextUtils.isEmpty(userId) || appContext == null) {
            return false;
        }
        File file = getDataFile();
        if (file == null || !file.exists()) {
            return false;
        }
        FileInputStream fis = null;
        try {
            fis = new FileInputStream(file);
            byte[] buf = new byte[(int) file.length()];
            int len = fis.read(buf);
            if (len <= 0) {
                return false;
            }
            String jsonStr = new String(buf, 0, len, "UTF-8");
            JSONObject obj = new JSONObject(jsonStr);
            
            reset();
            
            isEnable = obj.optBoolean("Enable", true);
            isAutoRegister = obj.optBoolean("AutoRegister", false);
            
            if (obj.has("Item")) {
                preRegisterVoiceprintItem = new AUIAICallVoiceprintItem(obj.getJSONObject("Item"));
            }
            if (obj.has("AutoItem")) {
                autoRegisterVoiceprintItem = new AUIAICallVoiceprintItem(obj.getJSONObject("AutoItem"));
            }
            
            Logger.i("Load preRegisterVoiceprintId: " + 
                    (preRegisterVoiceprintItem != null ? preRegisterVoiceprintItem.getVoiceprintId() : "未注册") +
                    " {uid: " + userId + ", pre: " + isPreEnv + "}");
            Logger.i("Load autoRegisterVoiceprintId: " + 
                    (autoRegisterVoiceprintItem != null ? autoRegisterVoiceprintItem.getVoiceprintId() : "未注册") +
                    " {uid: " + userId + ", pre: " + isPreEnv + "}");
            return true;
        } catch (Exception e) {
            Logger.e("Load data failed: " + e.getMessage());
        } finally {
            if (fis != null) {
                try {
                    fis.close();
                } catch (Exception ignore) {
                }
            }
        }
        return false;
    }

    /**
     * 保存声纹配置到 JSON 文件
     */
    private void saveData() {
        if (TextUtils.isEmpty(userId) || appContext == null) {
            return;
        }
        File file = getDataFile();
        if (file == null) {
            return;
        }
        FileOutputStream fos = null;
        try {
            JSONObject obj = new JSONObject();
            obj.put("Enable", isEnable);
            obj.put("AutoRegister", isAutoRegister);
            
            if (preRegisterVoiceprintItem != null) {
                obj.put("Item", preRegisterVoiceprintItem.toJson());
            }
            if (autoRegisterVoiceprintItem != null) {
                obj.put("AutoItem", autoRegisterVoiceprintItem.toJson());
            }
            
            fos = new FileOutputStream(file, false);
            fos.write(obj.toString().getBytes("UTF-8"));
            fos.flush();
            
            Logger.i("Save preRegisterVoiceprintId: " + 
                    (preRegisterVoiceprintItem != null ? preRegisterVoiceprintItem.getVoiceprintId() : "未注册") +
                    " {uid: " + userId + ", pre: " + isPreEnv + "}");
            Logger.i("Save autoRegisterVoiceprintId: " + 
                    (autoRegisterVoiceprintItem != null ? autoRegisterVoiceprintItem.getVoiceprintId() : "未注册") +
                    " {uid: " + userId + ", pre: " + isPreEnv + "}");
        } catch (Exception e) {
            Logger.e("Save data failed: " + e.getMessage());
        } finally {
            if (fos != null) {
                try {
                    fos.close();
                } catch (Exception ignore) {
                }
            }
        }
    }

    /**
     * 获取或创建 AppServerService
     */
    private ARTCAICallServiceImpl.AppServerService getAppServerService() {
        if (appServerService != null) {
            return appServerService;
        }
        // 使用 Manager 自身的 isPreEnv 状态，确保与数据文件路径保持一致
        String host = isPreEnv ? AUIAICallAgentDebug.PRE_HOST : AppServiceConst.HOST;
        appServerService = new ARTCAICallServiceImpl.AppServerService(host);
        Logger.i("Create AppServerService with host: " + host + " (isPreEnv: " + isPreEnv + ")");
        return appServerService;
    }

    /**
     * 声纹操作回调接口
     */
    public interface VoiceprintCallback {
        void onResult(boolean success, String errorMsg);
    }

    /**
     * 预注册声纹：录音 → OSS上传 → 服务端注册
     */
    public void startPreRegister(String wavFilePath, String authorization, VoiceprintCallback callback) {
        if (TextUtils.isEmpty(userId)) {
            if (callback != null) {
                callback.onResult(false, "userId is empty");
            }
            return;
        }

        // 第一步：获取 OSS 配置
        requestOssConfig(authorization, (ossConfig, errorMsg) -> {
            if (ossConfig == null) {
                if (callback != null) {
                    callback.onResult(false, errorMsg);
                }
                return;
            }

            // 第二步：上传到 OSS
            uploadFileToOss(wavFilePath, ossConfig, (ossUrl, error) -> {
                if (ossUrl == null) {
                    if (callback != null) {
                        callback.onResult(false, error);
                    }
                    return;
                }

                // 第三步：注册声纹
                registerVoiceprint(ossUrl, authorization, (voiceprintId, error1) -> {
                    if (voiceprintId == null) {
                        if (callback != null) {
                            callback.onResult(false, error1);
                        }
                        return;
                    }

                    // 第四步：保存到本地 JSON
                    AUIAICallVoiceprintItem item = new AUIAICallVoiceprintItem(voiceprintId);
                    item.setFilePath(wavFilePath);
                    item.setOssUrl(ossUrl);
                    preRegisterVoiceprintItem = item;
                    saveData();

                    if (callback != null) {
                        callback.onResult(true, null);
                    }
                });
            });
        });
    }

    /**
     * 请求 OSS 配置
     */
    private void requestOssConfig(String authorization, OssConfigCallback callback) {
        try {
            JSONObject body = new JSONObject();
            body.put("user_id", userId);
            getAppServerService().postAsync(
                    null,
                    "/api/v2/aiagent/getOssConfig",
                    authorization,
                    body,
                    new IARTCAICallService.IARTCAICallServiceCallback() {
                        @Override
                        public void onSuccess(JSONObject jsonObject) {
                            if (callback != null) {
                                callback.onResult(jsonObject, null);
                            }
                        }

                        @Override
                        public void onFail(int errorCode, String errorMsg) {
                            if (callback != null) {
                                callback.onResult(null, errorMsg);
                            }
                        }
                    }
            );
        } catch (JSONException e) {
            if (callback != null) {
                callback.onResult(null, e.getMessage());
            }
        }
    }

    /**
     * 上传文件到 OSS
     */
    private void uploadFileToOss(String wavFilePath, JSONObject ossConfig, OssUploadCallback callback) {
        try {
            String bucket = ossConfig.optString("bucket", "");
            String region = ossConfig.optString("region", "");
            String accessKeyId = ossConfig.optString("access_key_id", "");
            String accessKeySecret = ossConfig.optString("access_key_secret", "");
            String securityToken = ossConfig.optString("sts_token", "");
            String basePath = ossConfig.optString("base_path", "");

            String endpoint = "https://" + region + ".aliyuncs.com";
            String objectKey = basePath + "/" + new File(wavFilePath).getName();

            OSSCredentialProvider credentialProvider = new OSSStsTokenCredentialProvider(
                    accessKeyId, accessKeySecret, securityToken);
            OSSClient oss = new OSSClient(appContext, endpoint, credentialProvider);
            oss.setRegion(region);

            PutObjectRequest put = new PutObjectRequest(bucket, objectKey, wavFilePath);

            OSSAsyncTask task = oss.asyncPutObject(put, new OSSCompletedCallback<PutObjectRequest, PutObjectResult>() {
                @Override
                public void onSuccess(PutObjectRequest request, PutObjectResult result) {
                    try {
                        String ossUrl = oss.presignConstrainedObjectURL(bucket, objectKey, 3600);
                        if (callback != null) {
                            callback.onResult(ossUrl, null);
                        }
                    } catch (ClientException e) {
                        if (callback != null) {
                            callback.onResult(null, e.getMessage());
                        }
                    }
                }

                @Override
                public void onFailure(PutObjectRequest request, ClientException clientException, ServiceException serviceException) {
                    String errorMsg = "Unknown error";
                    if (clientException != null) {
                        errorMsg = clientException.getMessage();
                    } else if (serviceException != null) {
                        errorMsg = serviceException.getRawMessage();
                    }
                    if (callback != null) {
                        callback.onResult(null, errorMsg);
                    }
                }
            });
        } catch (Exception e) {
            if (callback != null) {
                callback.onResult(null, e.getMessage());
            }
        }
    }

    /**
     * 注册声纹
     */
    private void registerVoiceprint(String audioOssUrl, String authorization, VoiceprintIdCallback callback) {
        try {
            String voiceprintId = userId + "_" + System.currentTimeMillis();
            
            JSONObject input = new JSONObject();
            input.put("Type", "oss");
            input.put("Data", audioOssUrl);
            input.put("Format", "wav");

            JSONObject body = new JSONObject();
            body.put("user_id", userId);
            body.put("region", getRegion());
            body.put("voiceprint_id", voiceprintId);
            body.put("input", input.toString());

            getAppServerService().postAsync(
                    null,
                    "/api/v2/aiagent/setAIAgentVoiceprint",
                    authorization,
                    body,
                    new IARTCAICallService.IARTCAICallServiceCallback() {
                        @Override
                        public void onSuccess(JSONObject jsonObject) {
                            String id = jsonObject.optString("voiceprint_id", null);
                            if (TextUtils.isEmpty(id)) {
                                if (callback != null) {
                                    callback.onResult(null, jsonObject.toString());
                                }
                            } else {
                                if (callback != null) {
                                    callback.onResult(id, null);
                                }
                            }
                        }

                        @Override
                        public void onFail(int errorCode, String errorMsg) {
                            if (callback != null) {
                                callback.onResult(null, errorMsg);
                            }
                        }
                    }
            );
        } catch (JSONException e) {
            if (callback != null) {
                callback.onResult(null, e.getMessage());
            }
        }
    }

    /**
     * 生成声纹ID，用于无感注册模式
     */
    public String generateVoiceprintId() {
        currentVoiceprintId = userId + "_" + System.currentTimeMillis();
        Logger.i("Generated voiceprint ID for current call: " + currentVoiceprintId);
        return currentVoiceprintId;
    }

    /**
     * 获取当前通话使用的声纹 ID（可能是刚生成的临时 ID）
     */
    public String getCurrentVoiceprintId() {
        return currentVoiceprintId;
    }

    /**
     * 无感注册模式：声纹注册完成时调用
     */
    public void onAutoRegisted(String voiceprintId) {
        if (!isAutoRegister) {
            Logger.i("Skip onAutoRegisted: not in auto-register mode");
            return;
        }
        
        if (TextUtils.isEmpty(voiceprintId)) {
            Logger.w("Skip onAutoRegisted: voiceprintId is empty");
            return;
        }
        
        // 验证 ID 是否匹配当前通话生成的 ID
        if (currentVoiceprintId != null && !TextUtils.equals(currentVoiceprintId, voiceprintId)) {
            Logger.w("VoiceprintId mismatch! Expected: " + currentVoiceprintId + ", Got: " + voiceprintId);
        }
        
        if (autoRegisterVoiceprintItem != null) {
            if (TextUtils.equals(autoRegisterVoiceprintItem.getVoiceprintId(), voiceprintId)) {
                Logger.i("Auto-register voiceprint already saved: " + voiceprintId);
                currentVoiceprintId = null;  // 清除临时 ID
                return;
            } else {
                Logger.w("Auto register voiceprintId changed, old: " +
                        autoRegisterVoiceprintItem.getVoiceprintId() + ", new: " + voiceprintId);
            }
        }
        
        autoRegisterVoiceprintItem = new AUIAICallVoiceprintItem(voiceprintId);
        saveData();
        currentVoiceprintId = null;  // 清除临时 ID
        Logger.i("Auto-register voiceprint saved successfully: " + voiceprintId);
    }

    /**
     * 删除无感注册的声纹（乐观删除策略：先删除本地，再通知服务端）
     */
    public void removeAutoRegister(String authorization, VoiceprintCallback callback) {
        if (autoRegisterVoiceprintItem == null) {
            if (callback != null) {
                callback.onResult(true, null);
            }
            return;
        }

        final String voiceprintId = autoRegisterVoiceprintItem.getVoiceprintId();
        
        autoRegisterVoiceprintItem = null;
        saveData();
        Logger.i("Auto-register voiceprint removed locally: " + voiceprintId);
        
        // 立即回调成功，不等待服务端响应
        if (callback != null) {
            callback.onResult(true, null);
        }
        
        // 尝试通知服务端清除
        try {
            JSONObject body = new JSONObject();
            body.put("user_id", userId);
            body.put("region", getRegion());
            body.put("voiceprint_id", voiceprintId);
            body.put("registration_mode", "Implicit");

            Logger.i("Notifying server to clear auto-register voiceprint: " + voiceprintId);

            getAppServerService().postAsync(
                    null,
                    "/api/v2/aiagent/clearAIAgentVoiceprint",
                    authorization,
                    body,
                    new IARTCAICallService.IARTCAICallServiceCallback() {
                        @Override
                        public void onSuccess(JSONObject response) {
                            int code = response.optInt("code", -1);
                            String message = response.optString("message", "");
                            if (code == 200) {
                                Logger.i("Server cleared auto-register voiceprint successfully");
                            } else {
                                Logger.w("Server clear failed with code: " + code + ", message: " + message);
                            }
                        }

                        @Override
                        public void onFail(int errorCode, String errorMsg) {
                            // 服务端清除失败不影响用户体验，只记录日志
                            Logger.w("Server clear voiceprint failed (ignored): errorCode=" + errorCode + ", errorMsg=" + errorMsg);
                        }
                    }
            );
        } catch (JSONException e) {
            Logger.w("Failed to build server clear request (ignored): " + e.getMessage());
        }
    }

    /**
     * 删除预注册的声纹
     */
    public void removePreRegister(String authorization, VoiceprintCallback callback) {
        if (preRegisterVoiceprintItem == null) {
            if (callback != null) {
                callback.onResult(true, null);
            }
            return;
        }

        final String voiceprintId = preRegisterVoiceprintItem.getVoiceprintId();
        try {
            JSONObject body = new JSONObject();
            body.put("user_id", userId);
            body.put("region", getRegion());
            body.put("voiceprint_id", voiceprintId);
            body.put("registration_mode", "Explicit");

            Logger.i("Clearing pre-register voiceprint, voiceprintId: " + voiceprintId);

            getAppServerService().postAsync(
                    null,
                    "/api/v2/aiagent/clearAIAgentVoiceprint",
                    authorization,
                    body,
                    new IARTCAICallService.IARTCAICallServiceCallback() {
                        @Override
                        public void onSuccess(JSONObject response) {
                            int code = response.optInt("code", -1);
                            String message = response.optString("message", "");
                            Logger.i("clearAIAgentVoiceprint result: code=" + code + ", message=" + message);

                            if (code == 200) {
                                preRegisterVoiceprintItem = null;
                                saveData();
                                Logger.i("Pre-register voiceprint cleared successfully");
                                if (callback != null) {
                                    callback.onResult(true, null);
                                }
                            } else {
                                Logger.w("clearAIAgentVoiceprint returned non-200 code: " + code);
                                if (callback != null) {
                                    callback.onResult(false, message);
                                }
                            }
                        }

                        @Override
                        public void onFail(int errorCode, String errorMsg) {
                            Logger.e("clearAIAgentVoiceprint onFail: errorCode=" + errorCode + ", errorMsg=" + errorMsg);
                            if (callback != null) {
                                callback.onResult(false, errorMsg);
                            }
                        }
                    }
            );
        } catch (JSONException e) {
            Logger.e("removePreRegister exception: " + e.getMessage());
            if (callback != null) {
                callback.onResult(false, e.getMessage());
            }
        }
    }

    // 内部回调接口
    private interface OssConfigCallback {
        void onResult(JSONObject ossConfig, String error);
    }

    private interface OssUploadCallback {
        void onResult(String ossUrl, String error);
    }

    private interface VoiceprintIdCallback {
        void onResult(String voiceprintId, String error);
    }
}
